File: android/app/src/main/java/com/noop/push/ServerMetricOwnership.kt
package com.noop.push

import org.json.JSONObject

/** Rendering ownership is durable; each new value still needs current canonical authorization. */
data class ServerMetricOwnership(
    val project: String,
    val ownerId: String,
    val deviceId: String,
    val claims: Map<String, Claim> = emptyMap(),
    val ownedFamilies: Set<String> = emptySet(),
) {
    data class Claim(val metrics: Set<String>, val algorithmVersion: String, val inputRevision: Long,
                     val manifestHash: String?, val featureManifestHash: String?)

    val metrics: Set<String> get() = claims.values.flatMap { it.metrics }.toSet() +
        ownedFamilies.flatMap { ServerComputeContract.familyMetrics[it].orEmpty() }
    fun owns(metric: String) = metric in metrics

    fun presentation(cache: ServerScoreDayCache?, day: String, readFailed: Boolean = false): ServerScoreDayCache? {
        if (cache != null && cache.day == day && cache.ownerId.equals(ownerId, true) &&
            cache.features.values.all { it.deviceId == null || it.deviceId == deviceId } &&
            (cache.compute == null || cache.compute.project.trimEnd('/') == project.trimEnd('/') &&
                cache.compute.families.values.all { it.deviceId == deviceId ||
                    it.deviceId == null && it.reason == "device_registration_pending" && !it.authorized })) {
            return cache.copy(ownedMetrics = metrics + cache.compute?.ownedMetrics.orEmpty(), stale = cache.stale || readFailed,
                readFailure = if (readFailed) "server_read_failed" else cache.readFailure)
        }
        if (metrics.isEmpty()) return null
        return ServerScoreDayCache(day, "per_feature", null, emptyList(), null, false, 0,
            ownerId, features = claims.mapValues { (_, claim) ->
                ServerScoreFeatureCache("pending", "awaiting_server_result", deviceId,
                    claim.algorithmVersion, null, null, null, null, null, null, null)
            }, ownedMetrics = metrics, readFailure = if (readFailed) "server_read_failed" else null)
    }

    // Historical details, baselines and sessions do not have a complete canonical replacement yet.
    val canRetireDailyKernel: Boolean get() = metrics.containsAll(dailyKernelOutputs)

    fun observe(cache: ServerScoreDayCache): ServerMetricOwnership {
        if (!cache.ownerId.equals(ownerId, ignoreCase = true)) return this
        val next = claims.toMutableMap()
        cache.compute?.takeIf { it.project.trimEnd('/') == project.trimEnd('/') }?.families?.forEach { (key, family) ->
            val version = family.algorithmVersion
            val revision = family.inputRevision
            if (family.deviceId == deviceId && family.resultRevision != null && version != null && revision != null) {
                val prior = next[key]
                if (prior?.algorithmVersion != version || prior.inputRevision <= revision)
                    next[key] = Claim(family.metrics, version, revision, family.manifestHash, family.featureManifestHash)
            }
        }
        cache.features.forEach { (key, feature) ->
            if (cache.compute != null) return@forEach
            val version = feature.algorithmVersion
            val revision = feature.inputRevision
            if (key !in featureMetrics || feature.deviceId != deviceId || !feature.isCanonicalAvailable ||
                version == null || revision == null || revision < 0 || feature.computedAt == null) return@forEach
            val prior = next[key]
            if (prior?.algorithmVersion == version && prior.inputRevision > revision) return@forEach
            next[key] = Claim(featureMetrics.getValue(key), version, revision, feature.manifestHash, feature.featureManifestHash)
        }
        val families = cache.compute?.takeIf { it.project.trimEnd('/') == project.trimEnd('/') }
            ?.families?.filterValues { it.deviceId == deviceId ||
                it.deviceId == null && it.reason == "device_registration_pending" && !it.authorized }?.keys.orEmpty()
        return copy(claims = next, ownedFamilies = ownedFamilies + families)
    }

    fun encode(): String = JSONObject().put("project", project).put("ownerId", ownerId).put("deviceId", deviceId)
        .put("ownedFamilies", org.json.JSONArray(ownedFamilies.sorted()))
        .put("claims", JSONObject(claims.mapValues { (_, claim) -> JSONObject()
            .put("metrics", org.json.JSONArray(claim.metrics.sorted()))
            .put("algorithmVersion", claim.algorithmVersion).put("inputRevision", claim.inputRevision)
            .put("manifestHash", claim.manifestHash ?: JSONObject.NULL)
            .put("featureManifestHash", claim.featureManifestHash ?: JSONObject.NULL) })).toString()

    companion object {
        val dailyKernelOutputs = setOf(
            "hrv_rmssd_ms", "hrv_sdnn_ms", "resting_hr_bpm", "resp_rate_bpm", "recovery", "strain",
            "sleep_sessions", "sleep_total_min", "sleep_in_bed_min", "sleep_awake_min", "sleep_light_min",
            "sleep_deep_min", "sleep_rem_min", "sleep_efficiency", "sleep_performance", "disturbances",
            "steps", "active_kcal_est", "steps_est", "exercise_count", "workouts", "spo2_pct", "spo2_red",
            "spo2_ir", "skin_temp_c", "skin_temp_dev_c", "fitness_age", "vo2max_est", "vitality", "body_age",
            "sleep_need_min", "sleep_debt_min", "sleep_consistency", "historical_baselines",
        )
        val featureMetrics = mapOf(
            "hrv" to setOf("hrv_rmssd_ms", "resting_hr_bpm", "recovery", "strain", "spo2_pct", "skin_temp_c", "skin_temp_dev_c"),
            "respiration" to setOf("resp_rate_bpm"),
            "sleep" to setOf("sleep_performance", "sleep_total_min", "sleep_in_bed_min", "sleep_awake_min",
                "sleep_light_min", "sleep_deep_min", "sleep_rem_min", "sleep_efficiency", "sleep_sessions"),
        )

        fun restore(raw: String?, project: String, ownerId: String, deviceId: String): ServerMetricOwnership {
            val empty = ServerMetricOwnership(project.trimEnd('/'), ownerId.lowercase(), deviceId)
            return runCatching {
                val root = JSONObject(requireNotNull(raw))
                require(root.getString("project") == empty.project && root.getString("ownerId") == empty.ownerId &&
                    root.getString("deviceId") == empty.deviceId)
                val values = root.getJSONObject("claims")
                val owned = root.optJSONArray("ownedFamilies")?.let { array ->
                    (0 until array.length()).map { array.getString(it) }.toSet().also { require(ServerComputeContract.familyIDs.containsAll(it)) }
                }.orEmpty()
                empty.copy(ownedFamilies = owned, claims = values.keys().asSequence().filter { it in featureMetrics || it in ServerComputeContract.familyMetrics }.associateWith { key ->
                    val item = values.getJSONObject(key)
                    val metricArray = item.getJSONArray("metrics")
                    val metrics = (0 until metricArray.length()).map { metricArray.getString(it) }.toSet()
                    require((featureMetrics[key].orEmpty() + ServerComputeContract.familyMetrics[key].orEmpty()).containsAll(metrics))
                    Claim(metrics, item.getString("algorithmVersion").also { require(it.isNotEmpty()) }, item.getLong("inputRevision").also { require(it >= 0) },
                        item.optString("manifestHash").takeUnless { it.isEmpty() || it == "null" },
                        item.optString("featureManifestHash").takeUnless { it.isEmpty() || it == "null" })
                })
            }.getOrDefault(empty)
        }
    }
}

File: Packages/WhoopStore/Sources/WhoopStore/ServerMetricOwnership.swift
import Foundation

/// A rendering cutover survives empty days and failures. Qualification is checked again for
/// every value; retaining ownership never authorizes a previously revoked result.
public struct ServerMetricOwnership: Codable, Equatable, Sendable {
    public struct Scope: Codable, Equatable, Sendable {
        public let project: String
        public let ownerID: String
        public let deviceID: String

        public init(project: String, ownerID: String, deviceID: String) {
            self.project = project.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            self.ownerID = ownerID.lowercased()
            self.deviceID = deviceID
        }
    }

    public struct Claim: Codable, Equatable, Sendable {
        public let metrics: Set<String>
        public let algorithmVersion: String
        public let inputRevision: Int64
        public let manifestHash: String?
        public let featureManifestHash: String?
    }

    public let scope: Scope
    public private(set) var claims: [String: Claim] = [:]

    public init(scope: Scope) { self.scope = scope }

    // Only outputs actually serialized by the physiology read contract belong here. The
    // historical shadow contract's richer details cannot be inferred from these three features.
    public static let featureMetrics: [String: Set<String>] = [
        "hrv": ["hrv_rmssd_ms", "resting_hr_bpm", "recovery", "strain", "spo2_pct",
                "skin_temp_c", "skin_temp_dev_c"],
        "respiration": ["resp_rate_bpm"],
        "sleep": ["sleep_performance", "sleep_total_min", "sleep_in_bed_min", "sleep_awake_min",
                  "sleep_light_min", "sleep_deep_min", "sleep_rem_min", "sleep_efficiency", "sleep_sessions"],
    ]

    public var metrics: Set<String> {
        claims.values.reduce(into: Set<String>()) { $0.formUnion($1.metrics) }
    }

    @discardableResult
    public mutating func observe(_ cache: ServerScoreDayCache) -> Bool {
        guard cache.ownerId.lowercased() == scope.ownerID else { return false }
        let before = claims
        for (key, feature) in cache.features where Self.featureMetrics[key] != nil {
            guard feature.deviceId == scope.deviceID, feature.isCanonicalAvailable,
                  let version = feature.algorithmVersion, let revision = feature.inputRevision,
                  revision >= 0, feature.computedAt != nil else { continue }
            if let prior = claims[key], prior.algorithmVersion == version,
               prior.inputRevision > revision { continue }
            claims[key] = Claim(metrics: Self.featureMetrics[key] ?? [], algorithmVersion: version, inputRevision: revision,
                manifestHash: feature.manifestHash, featureManifestHash: feature.featureManifestHash)
        }
        return before != claims
    }

    public func owns(_ metric: String) -> Bool { metrics.contains(metric) }

    public func presentation(_ cache: ServerScoreDayCache?, day: String, readFailed: Bool = false) -> ServerScoreDayCache? {
        if var cache, cache.day == day, cache.ownerId.lowercased() == scope.ownerID,
           cache.features.values.allSatisfy({ $0.deviceId == nil || $0.deviceId == scope.deviceID }) {
            cache.ownedMetrics = metrics
            if readFailed { cache.stale = true; cache.readFailure = "server_read_failed" }
            return cache
        }
        guard !claims.isEmpty else { return nil }
        var pending = ServerScoreDayCache(day: day, algorithmVersion: "per_feature", daily: nil,
            nights: [], computedAt: nil, stale: false, fetchedAt: .distantPast)
        pending.ownerId = scope.ownerID
        pending.ownedMetrics = metrics
        pending.readFailure = readFailed ? "server_read_failed" : nil
        // This is a local presentation of an outstanding server read, never a cached publication.
        for (key, claim) in claims {
            pending.features[key] = ServerScoreFeatureCache(status: "pending", reason: "awaiting_server_result",
                deviceId: scope.deviceID, algorithmVersion: claim.algorithmVersion, inputRevision: nil,
                requiredRevision: nil, computedAt: nil, observedThrough: nil, publicationStatus: nil,
                archiveStatus: nil, manifestHash: nil)
        }
        return pending
    }

    /// A daily kernel also produces history, workouts, baselines and context. The narrow
    /// physiology response cannot prove that its complete replacement is consumed.
    public static let dailyKernelOutputs: Set<String> = [
        "hrv_rmssd_ms", "hrv_sdnn_ms", "resting_hr_bpm", "resp_rate_bpm", "recovery", "strain",
        "sleep_sessions", "sleep_total_min", "sleep_in_bed_min", "sleep_awake_min", "sleep_light_min",
        "sleep_deep_min", "sleep_rem_min", "sleep_efficiency", "sleep_performance", "disturbances",
        "steps", "active_kcal_est", "steps_est", "exercise_count", "workouts", "spo2_pct", "spo2_red",
        "spo2_ir", "skin_temp_c", "skin_temp_dev_c", "fitness_age", "vo2max_est", "vitality", "body_age",
        "sleep_need_min", "sleep_debt_min", "sleep_consistency", "historical_baselines",
    ]
    public var canRetireDailyKernel: Bool { Self.dailyKernelOutputs.isSubset(of: metrics) }
}

/// The scope is part of both the key and payload. A prior project's or wearable's cutover
/// cannot be inherited through a reused installation or the last day fetched.
public final class ServerMetricOwnershipStore {
    private let defaults: UserDefaults
    private static let lock = NSLock()

    public init(defaults: UserDefaults = .standard) { self.defaults = defaults }

    private func key(_ scope: ServerMetricOwnership.Scope) -> String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        return "noop.compute.ownership.v1." + (try! encoder.encode(scope)).base64EncodedString()
    }

    private func read(_ scope: ServerMetricOwnership.Scope) -> ServerMetricOwnership {
        guard let data = defaults.data(forKey: key(scope)),
              let value = try? JSONDecoder().decode(ServerMetricOwnership.self, from: data),
              value.scope == scope,
              value.claims.allSatisfy({ key, claim in
                  claim.inputRevision >= 0 && !claim.algorithmVersion.isEmpty &&
                    claim.metrics.isSubset(of: ServerMetricOwnership.featureMetrics[key] ?? [])
              }) else { return .init(scope: scope) }
        return value
    }

    public func load(_ scope: ServerMetricOwnership.Scope) -> ServerMetricOwnership {
        Self.lock.lock(); defer { Self.lock.unlock() }
        return read(scope)
    }

    public func observe(_ cache: ServerScoreDayCache, scope: ServerMetricOwnership.Scope) -> ServerMetricOwnership {
        Self.lock.lock(); defer { Self.lock.unlock() }
        var value = read(scope)
        if value.observe(cache), let data = try? JSONEncoder().encode(value) {
            defaults.set(data, forKey: key(scope))
        }
        return value
    }
}
